Update EIP-8025: Move to Draft (#11604)
ethereum/EIPs: EIPS/eip-8025.md

eip: 8025
title: Optional Execution Proofs
description: Introducing optional execution proofs on the consensus layer
author: Kevaundray Wedderburn (@kevaundray), Justin Drake (@JustinDrake) <justin@ethereum.org>, Ignacio Hagopian (@jsign), Han (@han0110), Francesco Risitano (@frisitano)
discussions-to: https://ethereum-magicians.org/t/eip-optional-execution-proofs/25500
status: Draft
type: Standards Track
category: Core
created: 2025-09-17

> ---

Reviewer: Review with me, but I believe the required EIPs for this one are:

EIP-4844, EIP-6110, EIP-7002, EIP-7251, EIP-7928

(would thus need a `required: ` tag)

Reviewer 2: Makes sense! I'll also add 7732 (ePBS) since it technically provides some extra proving time due to slot restructuring.

## Abstract

> Execution proofs enable consensus layer (CL) nodes to perform stateless validation of execution payloads. Altruistic _proof-generating nodes_ produce execution proofs and broadcast them over the CL p2p network. _Proof-verifying nodes_ consume gossiped execution proofs to assert the validity of execution payloads. Verification is _stateless_ and sublinear in the gas limit, so a validator's resource requirements scale sublinearly, an improvement on the linear cost of re-execution. The mechanism is fully opt-in and changes no consensus rules, allowing the proving stack to mature in production; a separate, future EIP could subsequently propose making execution proofs mandatory once they have matured.

Reviewer: As a complete outsider reading this, the proof is abstract and it is not clear that this is a zero-knowledge proof. This should be added.

Reviewer: Regarding the altruistic word used here (I understand the intention), is it also altruistic if these nodes would be at higher risk of slashing in case these optional proofs are invalid?
Why should I run one?
My current feeling is that these are "research nodes" and these are deployed on testnets, or maybe as nodes which are not staking but deployed on Mainnet.
If "research node" (node intended to do research on enabling zk proofs) is not the intended "idea" here this should be made clear somewhere.
Noting my general perspective or feeling towards the idea of this EIP, if this is not the right idea then this should be clarified, as otherwise other people reading this EIP might also fall into this "trap" or misconception, leading to confusion.

Reviewer 2: > As a complete outsider reading this, the proof is abstract, and it is not clear that this is a zero-knowledge proof. This should be added.

The proofs are not actually zero-knowledge, although this term is often abused out there. But in the EIP we don't mention is zero-knowledge because it isn't. 😄 

Do the "execution proofs" sound too abstract for the explanation? It should be understood only as a cryptographic proof that the execution of a program was correct, with a corresponding output. (Since it isn't zero-knowledge, we don't have to give any details about what this is).

Do you think it is useful to extend what an " execution proof " is?

Reviewer 2: > Regarding the altruistic word used here (I understand the intention), is it also altruistic if these nodes would be at higher risk of slashing in case these optional proofs are invalid? Why should I run one?

I think this would be altruistic too. I would say even more altruistic than without slashing since apart from not being paid it can also lose money. In both realities (i.e. with and without slashing), this EIP offers no incentives to generate proofs.

To put it differently: since the EIP doesn't propose prover incentives, anyone generating proofs is, in some way, altruistic, since it isn't rational to generate proofs.

I'll add some extra sentence to highlight that this EIP is not proposing any incentives thus it isn't rational to be a prover, which I think can help clarify this "altruistic" role.

## Motivation

Today, verifying a beacon block requires re-execution of its execution payload against the EL state — the full set of account and storage trie data. The cost of re-execution scales linearly with the gas limit. This couples a node's resource requirements to the state size and the chain's throughput.

This EIP introduces an opt-in path with the following properties:

> - **Sublinear verification cost.** Both verifier cost and proof size scale as `O(log² n)` in the size `n` of the witnessed computation. Since `n` is bounded by the payload's gas limit, verifier cost and proof size are correspondingly sublinear in the gas limit.

Reviewer: Nit: as someone not having much knowledge of zk computation, I do not know what "witnessed computation" here means. But the sentence makes it clear that the verifier cost and the proof size scale as O(log²n) in gas limit.
What is the verifier "cost"? Is this runtime?

Reviewer: I am also confused, because this seems to imply this O(log²n) is "the scale" here. I would assume this depends on the proof system (?).
I think this needs a more clear definition of what `n` is and also some intuition why O(log²n) is the bound here. Maybe this is a trivial knowledge in the ZK-world, but this makes me slightly uncomfortable 😅

Author: > What is the verifier "cost"? Is this runtime?
Correct, it is runtime. 

The "witnessed computation" is payload execution. I agree that having the term "witnessed computation" adds a layer of indirection and confusion, so maybe we can say something like "in the size `n` of the payload execution trace" or possibly "in the number of computation cycles `n` of payload execution". Happy for you to make a proposal on this.

`n` attempts to capture the analogous concept of CPU cycle count in the proof system domain. The concrete value `n` is a combination of guest program efficiency (consider this analogous to reth vs geth, for example) and zkVM efficiency (consider this analogous to ARM vs x86). 

> I am also confused, because this seems to imply this O(log²n) is "the scale" here. I would assume this depends on the proof system (?).

Correct, but we are currently only interested in STARK-based proof systems which follow this scale. If we want to be more generic, we can consider changing this scale to polylog(n) rather than log^2(n).

- **Stateless verification.** Verifying an execution proof requires only the proof itself and a public commitment to the payload — not the EL state.
- **Enhanced decentralization.** Sublinear, stateless verification keeps node operation tractable for participants with modest hardware and expertise, even as the gas limit and state grow.
- **Optional by design.** Deploying execution proofs as opt-in lets clients and operators build operational experience, collect performance metrics, and fine-tune the proving stack against live-network conditions without enshrining untested mechanics in the protocol or impacting validators that have not opted in.

> ## Specification

Reviewer: I think a terminology section before this one, or maybe before motivation, where terms for the non-zk-knowledgable-reader (like me) are defined would be incredibly helpful. This gives a super short introduction to the zk world. Terms:  prover, verifier, guest, host, proof node, proof engine, proof system, witness, public input.

Note: you can also link to a file under assets to direct a non familiar reader there.

Not sure here but I think it would be incredibly helpful for readers not familiar with ZK to get an intuitive definition of these (ZK) terms early in this EIP to be able to understand the EIP without using external resources

Reviewer: Esp. the terms private + public input and witness

Author: > I think a terminology section before this one, or maybe before motivation, where terms for the non-zk-knowledgable-reader (like me) are defined would be incredibly helpful. This gives a super short introduction to the zk world. Terms: prover, verifier, guest, host, proof node, proof engine, proof system, witness, public input.

Good idea. I agree this will be useful, I will add it.

Reviewer 2: @frisitano, since you're on a flight now, I'll add this `## Terminology` section in the other bundle of changes I'm making.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).

> This EIP defines two opt-in roles and the artefacts they exchange. Detailed specifications are maintained in the consensus-specs and execution-specs repositories seen below:

Reviewer: I think an illustrative diagram where we have prover, proof node, gossip, verifier in context of the "proof flow" would be helpful here for the reader.

Author: Good suggestion, I will add one.

Author: Added

    - [Guest program](https://github.com/ethereum/execution-specs/blob/85fc20ca5937719a854472a87cb48d01ef1dffca/src/ethereum/forks/amsterdam/stateless_guest.py)
    - [Host-side input assembly](https://github.com/ethereum/execution-specs/blob/85fc20ca5937719a854472a87cb48d01ef1dffca/src/ethereum/forks/amsterdam/stateless_host.py)
    - [Execution witness construction](https://github.com/ethereum/execution-specs/blob/85fc20ca5937719a854472a87cb48d01ef1dffca/src/ethereum/forks/amsterdam/stateless_host_exec_witness.py)
    - [SSZ serialization schema](https://github.com/ethereum/execution-specs/blob/85fc20ca5937719a854472a87cb48d01ef1dffca/src/ethereum/forks/amsterdam/stateless_ssz.py)
    - [Conformance tests](https://github.com/ethereum/execution-specs/tree/85fc20ca5937719a854472a87cb48d01ef1dffca/tests/amsterdam/eip8025_optional_proofs)

### Consensus Layer

This EIP introduces a new `execution_proof` gossip topic on the
consensus-layer p2p network carrying `SignedExecutionProof` messages,
and a `ProofEngine` interface that beacon nodes use to generate and
verify execution proofs. Proof generation and

> verification are delegated to an external proof node; the `ProofEngine`

Reviewer: What is a proof node? Is this a validator? Definition can be inferred from context, but to avoid guesswork, define it 😄 👍

Reviewer 2: I'll add the Terminology section, so I think will be addressed there 👍

is the in-client component that mediates communication with it. A node
MAY enable either or both of:

- **Proof-generating mode.** An active validator that opts in to act as
  a _prover_ — produces execution proofs and broadcasts them on the
  `execution_proof` gossip topic.
- **Proof-verifying mode.** A node that consumes gossiped proofs and
  treats a verified proof as a supplementary attestation of the
  corresponding payload's validity. Proof verification does not require local

>   EL state.

Reviewer: I assume that proof-verifying mode nodes will not re-execute and instead "only" do the proof verification? Or is in this initial step the idea that we do both? I guess that makes no sense to do both because the proof is essentially the "receipt" of the re-execution, so you would essentially do the job at least twice 😅

Author: >  Or is in this initial step the idea that we do both?

The intention is to do both initially. The idea is to deploy this at network scale (expecting approx. 10% of nodes to adopt it), but it should not be load-bearing on consensus or fork choice. This gives us an opportunity to collect data, optimise and build confidence in the new stack.

> At a high level, the proof lifecycle is:

Reviewer: Nit: a diagram could also be added here to illustrate the lifecycle

1. A prover reacts to a new beacon block, constructs the corresponding
   `NewPayloadRequest`, and calls
   `proof_engine.request_proofs(new_payload_request, ProofAttributes(proof_types))`
   to initiate proof generation for its configured `proof_types` on the proof node.
2. The proof node produces an `ExecutionProof` for each requested type.
   The prover signs each proof using the `DOMAIN_EXECUTION_PROOF` domain, wraps it in
   a `SignedExecutionProof`, and broadcasts it on the `execution_proof`
   topic.
3. Proof-verifying peers receive the gossip message, run the gossip
   validation rules, then call `proof_engine.verify_execution_proof` on
   the inner `ExecutionProof` to confirm payload validity.

#### Proof types and containers

The consensus layer introduces a small set of new types and constants to
identify proofs, bound their size, and route gossip signatures:

| Name                               | Value                      |
| ---------------------------------- | -------------------------- |
| `ProofType`                        | `uint8`                    |

> | `MAX_PROOF_SIZE`                   | `409600` (= 400 KiB)       |

Reviewer: Why 400 KiB? Is this a security parameter? Why specifically 400 KiB? Does this limit specific systems? For instance systems producing proofs faster, but which are also larger?

Author: This is set based on targeted bandwidth requirements for nodes. Should we elaborate/explain this rationale?

Author: Added rationale for this in the EIP

| `MAX_EXECUTION_PROOFS_PER_PAYLOAD` | `uint64(4)`                |
| `DOMAIN_EXECUTION_PROOF`           | `DomainType('0x0D000000')` |

Peers advertise their supported proof types via
`ExecutionProofStatus` (see [Req/resp domain](#reqresp-domain)), so new

> proof systems can be added without a hard fork or coordinated client

Reviewer: How does the coordination of these proof types work? 

A hard fork is not necessary by definition (?) because this EIP is opt-in?

Author: Nodes will advertise which proof types they support by specifying them in the `proof_types` field of the `ExecutionProofStatus` message. Peers will use this to determine which peers are useful to connect to and to make requests against. In practice, some out-of-protocol coordination will be required to iterate on supported proof types, but ultimately it's at the discretion of node operators to configure the proof types they support when starting their node. Correct, no hard fork necessary, as this is opt-in.

release.

The top-level gossip message is the `SignedExecutionProof`:

```python
class SignedExecutionProof(Container):
    message: ExecutionProof
    validator_index: ValidatorIndex
    signature: BLSSignature
```

Its inner `ExecutionProof` couples a serialized proof blob to the proof

> system and guest program it was generated against, and to the public

Reviewer: Nit: for people not aware of the zk terminology: a small definition of what a guest program is could help here.

Reviewer 2: Will be added in Terminology section 👍

payload execution to the EL client. The
interface comprises four operations and a `ProofAttributes` type:

```python
def verify_execution_proof(
    self: ProofEngine,
    execution_proof: ExecutionProof,
) -> bool: ...

def notify_new_payload(
    self: ProofEngine,
    new_payload_request: NewPayloadRequest,
) -> None: ...

def notify_forkchoice_updated(
    self: ProofEngine,
    head_block_hash: Hash32,
    safe_block_hash: Hash32,
    finalized_block_hash: Hash32,
) -> None: ...

@dataclass
class ProofAttributes:
    proof_types: Sequence[ProofType]

def request_proofs(
    self: ProofEngine,
    new_payload_request: NewPayloadRequest,
    proof_attributes: ProofAttributes,
) -> Root: ...
```

These operations decompose into three concerns:

- **Verification** (`verify_execution_proof`) is called on every incoming
  proof that passes gossip validation. The engine performs the
  proof-system-specific cryptographic verification and binds the result
  to the specific payload identified by
  `proof.public_input.new_payload_request_root`. A payload is considered
  proof-verified once `k` valid `SignedExecutionProof`s for it have been

>   verified (`k` is TBD <!-- TODO: pin a value of `k` before Review -->). Peers exchange their current

Reviewer: nit: TODO. Not sure if this is a TODO before this review or before the Review state of the EIP.

Author: The intention is that this should be decided before transitioning the state of this EIP to review.

Author: I made this clearer in the specs.

current state, verifies the BLS signature under `DOMAIN_EXECUTION_PROOF`,
and delegates cryptographic verification to the proof engine:

```python
def process_execution_proof(
    state: BeaconState,
    signed_proof: SignedExecutionProof,
    proof_engine: ProofEngine,
) -> None:
    proof_message = signed_proof.message

    # Verify prover is an active validator
    validator = state.validators[signed_proof.validator_index]
    assert is_active_validator(validator, get_current_epoch(state))

    domain = get_domain(state, DOMAIN_EXECUTION_PROOF, compute_epoch_at_slot(state.slot))
    signing_root = compute_signing_root(proof_message, domain)
    assert bls.Verify(validator.pubkey, signing_root, signed_proof.signature)

    # Verify the execution proof
    assert proof_engine.verify_execution_proof(proof_message)
```

`process_execution_proof` is invoked outside the beacon-block state-transition
function and produces no on-chain state change: it is the operational hook
that lets a node decide whether to treat a gossiped proof as a valid
attestation of an execution payload. Proof storage is managed by the
`ProofEngine`; retention is bounded by `proof_serve_range` (see

> [Req/resp domain](#reqresp-domain)): clients MUST retain proofs for

Reviewer: Nit: this MUST is I think rather important and should maybe be made clear earlier, this feels like a rather detailed part of the EIP

Author: I added this to the main "Consensus Layer" section.

canonical-chain blocks
back to the finalized checkpoint so they can serve `ExecutionProofsByRange`
and `ExecutionProofsByRoot` requests.

#### Proof gossip

Execution proofs are gossiped over a new global topic, `execution_proof`,
carrying `SignedExecutionProof` messages. Validation rules MUST pass before
forwarding. With `proof = signed_execution_proof.message`:

> - _[IGNORE]_ the proof's `new_payload_request_root` has been seen via

Reviewer: Nit: define that these IGNORE/REJECTs are intended to be gossipsub

Author: Made explicit in the text.

`[IGNORE]` rules bound steady-state bandwidth — once a node has a valid
proof for a given `(new_payload_request_root, proof_type)`, additional
proofs for the same tuple are dropped silently, regardless of who signs
them.

#### Req/resp domain

Three req/resp protocols let peers backfill, target, and coordinate proof
state:

- **`ExecutionProofsByRange`** — request `(start_slot, count, proof_types)`,
  response `List[SignedExecutionProof]`. Mirrors `BlobSidecarsByRange` and
  `DataColumnSidecarsByRange`: range-based, root-unaware, filtered by proof
  type. Used for bulk backfill of proofs after a syncing window.
- **`ExecutionProofsByRoot`** — request `List[ProofByRootIdentifier]` where
  each identifier carries `(block_root, List[proof_type])`, response
  `List[SignedExecutionProof]`. Used for targeted retrieval when a node
  knows the specific block roots whose proofs it is missing.
- **`ExecutionProofStatus`** — request and response share the same
  container: `(block_root, slot, proof_types)`. Peers exchange the most
  recent block they consider proof-verified, plus their dynamically
  advertised set of supported proof types. The dialing peer MUST send a

>   status on first connecting to any proof-aware peer, mirroring the Status

Reviewer: Nit: what is a proof-aware peer? Is this a peer implementing this EIP?

Author: Correct, the wording should be made clearer

Reviewer 2: I'll add to terminology section.

>   v2 handshake in Fulu.

Reviewer: Not sure if this specific Fulu definition is worth mentioning here 🤔 
If required, please link to it (this should be in consensus-specs (?) right?)

Author: I think it helps give protocol folks an analogy they are already familiar with and can reason about. I've added reference but happy to remove if preferable.

Clients MUST serve both range and root requests for any block on the
canonical chain whose slot is in `proof_serve_range` — defined as
`[finalized_checkpoint.slot, current_slot]`. Peers unable to reply respond
with `3: ResourceUnavailable` and MAY be descored. The shape of these
protocols composes with existing sync flows: a node syncing forward over
blocks pulls proofs by range; a node patching a specific gap pulls them by
root.

#### Discovery

> Discovery is extended with a new ENR key, `eproof`, encoded as a `uint8`. A

Reviewer: ```suggestion
Discovery is extended with a new Ethereum Node Record (ENR) key, `eproof`, encoded as a `uint8`. A
```

node is considered proof-aware if the field is present and non-zero,
allowing other peers to filter for proof-aware peers during discovery. This
mirrors the existing pattern used to advertise data-column awareness in

> Fulu and is intentionally cheap: no fork or fork-version bump is required

Reviewer: Nit: (ref to the specific advertised data in Fulu here)

for a node to start (or stop) advertising proof support.

#### Prover lifecycle

A prover's flow, as detailed in the prover guide referenced above, is
fully event-driven:

> 1. **Subscribe to two SSE event streams** at startup:

Reviewer: SSE is not defined, what is it?

Author: Server-side events are an HTTP-based events protocol. I'm not sure whether this surfaces an implementation detail that is too low-level for this EIP, or, if not, we can add it as a definition in the proposed terminology section?

to bind the proof to the payload request being checked.

A proof-verifying node accepts the execution-layer result only if:

- the proof verifies for the expected guest program;
- `successful_validation` is `true`;
- `new_payload_request_root` equals the SSZ `hash_tree_root` of the
  `NewPayloadRequest` associated with the payload being validated; and
- `chain_config` matches the chain ID and fork configuration expected by the
  verifier.

At a high level, proof generation follows this flow:

1. A host, usually a stateful execution-layer client, executes or constructs
   the block while recording the account, storage, code, and ancestor-header
   data used during execution.
2. The host builds an `ExecutionWitness`, wraps it with the Engine API
   `NewPayloadRequest`, chain configuration, and transaction public keys into a
   `StatelessInput`, and serializes that input as schema-prefixed SSZ bytes.
3. A prover runs the guest program on the serialized `StatelessInput`. The
   guest validates the chain configuration and witness scaffolding, executes the
   payload against witness-backed state, and returns `StatelessValidationResult`.

#### Stateless input and output

> The top-level guest input is private prover input:

Reviewer: Nit: (this is not understandable for a non-zk familiar person without the terminology section or doing an external lookup of these terms)

Reviewer 2: Will be added to the terminology section.

```python
class StatelessInput:
    new_payload_request: NewPayloadRequest
    witness: ExecutionWitness
    chain_config: ChainConfig
    public_keys: Tuple[Bytes, ...]
```

The guest output is public:

```python
class StatelessValidationResult:
    new_payload_request_root: Hash32
    successful_validation: bool
    chain_config: ChainConfig
```

Note this struct is the EL-counterpart of `PublicInput` described in the _Proof types and containers_
section.

At this level:

- `new_payload_request` is the Engine API payload data supplied by the
  consensus layer.
- `witness` is the account, storage, code, and ancestor-header data needed to
  execute the payload without local EL state.
- `chain_config` tells the guest which chain ID and fork configuration to
  validate against.

> - `public_keys` contains 65-byte uncompressed SEC1 transaction public keys in

Reviewer: SEC1 is not defined, what is this? secp256k1?

Reviewer 2: Ah, SEC1 refers to how the public key is encoded into bytes.
I'm doubting a bit about what to do here. Although SEC1 is a standard, I'm not sure how to make the EIP self-contained to describe how this works (mainly `0x04 || x || y`).

I think I'll reword to say that it contains the bytes and leave these details for someone trying to implement the EIP thus look at the spec where this is fully defined. Not sure talking about the encoding of elliptic curve points (although standards) is adding much value for EIP readers.

validation, and serializes the result.

```python
def run_stateless_guest(input_bytes: Bytes) -> Bytes:
    stateless_input = deserialize_stateless_input(input_bytes)
    stateless_output = verify_stateless_new_payload(stateless_input)
    return serialize_stateless_output(stateless_output)
```

The main validation path first computes the public payload-request commitment.
It then validates fork configuration, checks the header witness, creates a
`WitnessState`, and delegates payload execution to the same `new_payload` path
used by the execution engine:

```python
def compute_new_payload_request_root(
    stateless_input: StatelessInput,
) -> Hash32:
    ssz_npr = _new_payload_request_to_ssz(stateless_input.new_payload_request)
    return Hash32(ssz_npr.hash_tree_root())

def validate_chain_config(
    chain_config: ChainConfig,
    new_payload_request: NewPayloadRequest,
) -> ForkConfig:
    active_fork = chain_config.active_fork
    execution_payload = new_payload_request.execution_payload

    if not _is_activation_active(active_fork.activation, execution_payload):
        raise InactiveForkConfigError(...)

>     if active_fork.fork != ProtocolFork.Amsterdam:

Reviewer: Nit: EIP only works on Amsterdam.

Reviewer 2: This is actually mostly a side-effect of the `execution-specs` architecture, since it has one spc impl per fork.

In the `Amsterdam` spec, the guest program can simply work only in Amsterdam. But when we have Hegota, it will also work there.

I agree this `execution-specs` reality is leaking a bit into this EIP code snippet.

I'll edit this code a bit to avodi this confusion.

eip: 8025
title: Optional Execution Proofs
description: Introducing optional execution proofs on the consensus layer

> author: Kevaundray Wedderburn (@kevaundray), Justin Drake (@JustinDrake), Ignacio Hagopian (@jsign), Han (@han0110), Francesco Risitano (@frisitano)

Reviewer: Justin's email is removed, is this intended?

Reviewer 2: I think so, but @frisitano might confirm. I think is just because we use GH username for everyone.

Author: I will review how Justin is typically referenced in eips and align

discussions-to: https://ethereum-magicians.org/t/eip-optional-execution-proofs/25500
status: Draft
type: Standards Track
category: Core
created: 2025-09-17

> requires: 4844, 6110, 7002, 7251, 7732, 7928

Reviewer: These EIPs are not referenced here by explicitly mentioning, so for instance when talking about blobs it is nice to refer directly to the EIP, instead of having the reader to guess what concept / EIP it references

---

## Abstract

Execution proofs enable consensus layer (CL) nodes to perform stateless validation of execution payloads. Altruistic _proof-generating nodes_ produce execution proofs and broadcast them over the CL p2p network. _Proof-verifying nodes_ consume gossiped execution proofs to assert the validity of execution payloads. Verification is _stateless_ and sublinear in the gas limit, so a validator's resource requirements scale sublinearly, an improvement on the linear cost of re-execution. The mechanism is fully opt-in and changes no consensus rules, allowing the proving stack to mature in production; a separate, future EIP could subsequently propose making execution proofs mandatory once they have matured. This EIP does not introduce incentives for proof-generating nodes; for that reason, they are considered altruistic and the mechanism is opt-in.

## Motivation

> Today, verifying a beacon block requires re-execution of its execution payload against the EL state — the full set of account and storage trie data. The cost of re-execution scales linearly with the gas limit. This couples a node's resource requirements to the state size and the chain's throughput.

Reviewer: consensus layer (CL) is defined in the abstract, but EL is used here without definition, so suggestion: 
```suggestion
Today, verifying a beacon block requires re-execution of its execution payload against the execution layer (EL) state — the full set of account and storage trie data. The cost of re-execution scales linearly with the gas limit. This couples a node's resource requirements to the state size and the chain's throughput.
```

  treats a verified proof as a supplementary attestation of the
  corresponding payload's validity. Proof verification does not require local
  EL state. Proof-verifying nodes still re-execute payloads themselves;
  verified proofs are an additional signal, not a replacement for
  re-execution. This separation lets clients deploy and measure the
  proving stack against live-network conditions without making
  fork-choice dependent on it. A subsequent EIP could propose promoting
  verified proofs to a load-bearing role and dropping re-execution once
  the stack has matured.

The following diagram summarises how proofs move between the two roles:

![Execution-proof flow between proof-generating and proof-verifying nodes via the `execution_proof` gossip topic](../assets/eip-8025/proof-flow.png)

At a high level, the proof lifecycle is:

1. A prover reacts to a new beacon block, constructs the corresponding
   `NewPayloadRequest`, and calls
   `proof_engine.request_proofs(new_payload_request, ProofAttributes(proof_types))`
   to initiate proof generation for its configured `proof_types` on the proof node.
2. The proof node produces an `ExecutionProof` for each requested type.

>    The prover signs each proof using the `DOMAIN_EXECUTION_PROOF` domain, wraps it in

Reviewer: `DOMAIN_EXECUTION_PROOF` is used here, but defined later (line 133)

Reviewer 2: Maybe this is okay since this provides a high-level description? Trying to pull the table from 133 upwards maybe already starts too low level.

I can think of two things:
1. Add here `(defined later)`
2. Remove `DOMAIN_EXECUTION_PROOF` reference, and simply says `The prover signs each proof,`

I think 2. should be okay -- the "domain" separation feels like a technical detail for a high-level explanation.

def build_stateless_input(
    block: Block,
    *,
    execution_witness: ExecutionWitness,
    execution_requests: ExecutionRequests,
    block_access_list: BlockAccessList,
    chain_id: U64,
) -> StatelessInput:
    ...

    new_payload = NewPayloadRequest(
        execution_payload=payload,
        versioned_hashes=tuple(versioned_hashes),
        parent_beacon_block_root=header.parent_beacon_block_root,
        execution_requests=execution_requests,
    )

    return StatelessInput(
        new_payload_request=new_payload,
        witness=execution_witness,
        chain_config=build_chain_config(chain_id),
        public_keys=tuple(public_keys),
    )
```

The host also provides the fork configuration corresponding to the payload being
validated:

```python
def build_chain_config(chain_id: U64) -> ChainConfig:
    return ChainConfig(
        chain_id=chain_id,
        active_fork=ForkConfig(
            fork=ProtocolFork.Amsterdam,
            activation=ForkActivation(
                block_number=None,
                timestamp=U64(0),
            ),
            blob_schedule=BlobSchedule(
                target=BLOB_SCHEDULE_TARGET,
                max=BLOB_SCHEDULE_MAX,

>                 base_fee_update_fraction=U64(BLOB_BASE_FEE_UPDATE_FRACTION),

Reviewer: For clarity, add that these are the constants from https://eips.ethereum.org/EIPS/eip-7892 (right?)

Reviewer 2: Correct. Nothing here is intended to be new.

    _apply_storage_writes(incr_storage_mpts, block_state.storage_writes)

    all_dirty_accounts = _get_all_dirty_accounts(block_state)
    _capture_pre_state_account_nodes(
        incr_account_mpt,
        block_state.account_reads,
        all_dirty_accounts,
    )
    _apply_account_writes(
        incr_account_mpt,
        incr_storage_mpts,
        block_state,
        all_dirty_accounts,
    )

    assert mpt_root(incr_account_mpt) == expected_post_state_root

    accessed_nodes = _collect_accessed_nodes(
        incr_account_mpt, incr_storage_mpts
    )

    return ExecutionWitness(
        state=tuple(sorted(accessed_nodes.values())),
        codes=tuple(codes),
        headers=tuple(ancestor_headers),
    )
```

#### SSZ encoding

The host serializes a `StatelessInput` by prefixing the SSZ bytes with
`STATELESS_INPUT_SCHEMA_ID`. This lets future guest programs select the correct
input schema as forks or encoding rules evolve.

```python
STATELESS_INPUT_SCHEMA_ID = 0x0001
STATELESS_INPUT_SCHEMA_ID_SIZE = 2

def serialize_stateless_input(
    stateless_input: StatelessInput,
) -> Bytes:
    ssz_obj = stateless_input_to_ssz(stateless_input)
    return Bytes(

>         STATELESS_INPUT_SCHEMA_ID_BYTES + bytes(ssz_obj.encode_bytes())

Reviewer: This constant `STATELESS_INPUT_SCHEMA_ID_BYTES` is not defined

Reviewer 2: @frisitano, do you mind adding:
```
STATELESS_INPUT_SCHEMA_ID_BYTES = STATELESS_INPUT_SCHEMA_ID.to_bytes(
    STATELESS_INPUT_SCHEMA_ID_SIZE,
    "big",
)
```
Into L941?

    )

def deserialize_stateless_input(data: Bytes) -> StatelessInput:
    if len(data) < STATELESS_INPUT_SCHEMA_ID_SIZE:
        raise ValueError("Stateless input is missing schema id")
    schema_id = int.from_bytes(
        data[:STATELESS_INPUT_SCHEMA_ID_SIZE],
        "big",
    )
    if schema_id != STATELESS_INPUT_SCHEMA_ID:
        raise ValueError(
            f"Unsupported stateless input schema id: 0x{schema_id:04x}"
        )
    ssz_obj = SszStatelessInput.decode_bytes(
        data[STATELESS_INPUT_SCHEMA_ID_SIZE:]
    )
    return ssz_to_stateless_input(ssz_obj)
```

The SSZ schema mirrors the dataclasses and applies bounded-list limits to the
wire representation:

```python
class SszExecutionWitness(Container):
    state: SszList[ByteList[MAX_BYTES_PER_WITNESS_NODE], MAX_WITNESS_NODES]
    codes: SszList[ByteList[MAX_BYTES_PER_CODE], MAX_WITNESS_CODES]

>     headers: SszList[ByteList[MAX_BYTES_PER_HEADER], MAX_WITNESS_HEADERS]

Reviewer: The constants used here are defined later, not a super big thing but consider moving the table before the usage here.

Reviewer 2: @frisitano, sounds good to me to move table in L1084 before this snippet.

eip: 8025
title: Optional Execution Proofs
description: Introducing optional execution proofs on the consensus layer
author: Kevaundray Wedderburn (@kevaundray), Justin Drake (@JustinDrake) <justin@ethereum.org>, Ignacio Hagopian (@jsign), Han (@han0110), Francesco Risitano (@frisitano)
discussions-to: https://ethereum-magicians.org/t/eip-optional-execution-proofs/25500
status: Draft
type: Standards Track
category: Core
created: 2025-09-17

> requires: 4844, 6110, 7002, 7251, 7732, 7928

Reviewer: ```suggestion
requires: 4844, 6110, 7002, 7251, 7732, 7892, 7928
```

| `MAX_PROOF_SIZE`                   | `409600` (= 400 KiB)       |
| `MAX_EXECUTION_PROOFS_PER_PAYLOAD` | `uint64(4)`                |
| `DOMAIN_EXECUTION_PROOF`           | `DomainType('0x0D000000')` |

`MAX_PROOF_SIZE` is set at `400 KiB` as a network-bandwidth budget: at the
gossip target of one proof per `(payload, proof_type)` and the cap of
`MAX_EXECUTION_PROOFS_PER_PAYLOAD = 4` proof types per payload, the
worst-case in-flight proof bandwidth per slot is `4 × 400 KiB = 1.6 MiB`
— of a similar order to existing per-slot blob-sidecar traffic. This
bounds steady-state traffic. Provers using systems that naturally
produce larger proofs are expected to apply proof compression
(recursion, succinct wrappers) before gossiping.

Peers advertise their supported proof types via
`ExecutionProofStatus` (see [Req/resp domain](#reqresp-domain)). The set
of `proof_types` a node speaks is a per-node, dynamic configuration; new
proof systems can be added by socializing a new `ProofType` value
out-of-band and rolling it out to consenting operators. Because this EIP
is opt-in and changes no consensus rules, adding, deprecating, or

> swapping proof systems does **not** require a hard fork or coordinated

Reviewer: ```suggestion
swapping proof systems does **not** require a hardfork or coordinated
```

Consistency, `hardfork` used elsewhere

  `notify_forkchoice_updated`) keep the engine in sync with the state
  of the beacon node. `notify_new_payload` lets the engine associate
  incoming proofs with the payloads they certify;
  `notify_forkchoice_updated` lets it track the canonical chain to
  drive proof retention and pruning.
- **Generation** (`request_proofs`) is used by proof-generating nodes to
  trigger asynchronous proof generation for a given payload and set of
  proof types. The return value is `new_payload_request.hash_tree_root()`,
  which the prover uses as a request id to correlate the proof-completion
  SSE event the proof node emits when a proof is ready to be fetched.

#### Beacon-chain state transition

EIP-8025 extends `process_block` to forward execution payload processing to
both the execution engine and the proof engine, and introduces a new
`process_execution_proof` handler for incoming proofs.

`process_block` is modified to pass `PROOF_ENGINE` to
`process_execution_payload`:

```python
def process_block(state: BeaconState, block: BeaconBlock) -> None:
    process_block_header(state, block)
    process_withdrawals(state, block.body.execution_payload)
    # [Modified in EIP8025]

>     process_execution_payload(state, block.body, EXECUTION_ENGINE, PROOF_ENGINE)

Reviewer: Nit: `PROOF_ENGINE` is not defined. I think it would be beneficial to define earlier what is meant by this.

    )

    return StatelessInput(
        new_payload_request=new_payload,
        witness=execution_witness,
        chain_config=build_chain_config(chain_id),
        public_keys=tuple(public_keys),
    )
```

The host also provides the fork configuration corresponding to the payload being
validated:

```python
def build_chain_config(chain_id: U64) -> ChainConfig:
    return ChainConfig(
        chain_id=chain_id,
        active_fork=ForkConfig(
            fork=ProtocolFork.Amsterdam,
            activation=ForkActivation(
                block_number=None,
                timestamp=U64(0),
            ),
            blob_schedule=BlobSchedule(
                target=BLOB_SCHEDULE_TARGET,
                max=BLOB_SCHEDULE_MAX,
                base_fee_update_fraction=U64(BLOB_BASE_FEE_UPDATE_FRACTION),
            ),
        ),
    )
```

The blob schedule fields and their `target`, `max`, and
`base_fee_update_fraction` parameters are defined by
[EIP-7892](./eip-7892.md); this EIP introduces no new blob-schedule
parameters.

The execution witness is built from the block-level read/write tracker and
pre-state trie data. The read/write tracker is a component already used in ELs

> to track which data is touched during execution to assist for block access list (BALs)

Reviewer: ```suggestion
to track which data is touched during execution to assist for block access list (BAL)
```

or:

```suggestion
to track which data is touched during execution to assist for block access lists (BALs)
```

    new_payload_request_root: Bytes32
    successful_validation: boolean
    chain_config: SszChainConfig
```

## Rationale

### Path to low-resource validation

This step moves the network measurably closer to low-resource validation. Stateless, constant-time payload verification breaks the coupling between a node's hardware requirements and the gas limit or state size, lowering the floor for who can meaningfully run a node and improving decentralisation.

### Build operational experience

The central design choice is to deploy execution proofs as an opt-in feature. Stateless validation, the witness format, and the proof system itself are maturing, but we need operational experience before making execution proofs load-bearing. Treating execution proofs as a non-critical artefact lets the stack mature on a live network — proof sizes, generation latency, verifier throughput, gossip behaviour, prover diversity — without placing any of that on the path of fork choice or attestation. A bug, outage, or poor parameter choice is contained to the nodes that opted in; it cannot fork the chain or affect validators that did not subscribe.

## Backwards Compatibility

> This EIP is fully opt-in and does not require a hardfork. No consensus rule changes: validators that do not enable either mode see no change to their behaviour, their bandwidth, or their attestation duties. Nodes that opt in additionally subscribe to the proof gossip topic, advertise themselves via the `eproof` ENR field, and may run a prover; this affects only their local resource profile.

Reviewer 2: @frisitano, not sure saying "does not require a hardfork" is somewhat aligned with what we're trying to do with proposing 8025 in the ACD process; so I feel this might be confusing.

I think we agree on the original intention; mostly trying to think if this is in conflict with our argument of fork scheduling as a resource allocation ask.

Author: good call - I'll change the wording here.

> ## Test Cases

Reviewer: Please add a short introductionary line, like:

```suggestion
## Test Cases

Conformance tests are maintained in the canonical spec repositories:
```

- [Execution Layer](https://github.com/ethereum/execution-specs/tree/85fc20ca5937719a854472a87cb48d01ef1dffca/tests/amsterdam/eip8025_optional_proofs)
- [Consensus Layer](https://github.com/ethereum/consensus-specs/tree/031e521f0a82b0e475ecfc13f79dd251a6284fc2/specs/_features/eip8025)

## Reference Implementation

- [Execution Layer](https://github.com/ethereum/execution-specs/tree/85fc20ca5937719a854472a87cb48d01ef1dffca/src/ethereum/forks/amsterdam)

## Security Considerations

**Gossip surface.** Proofs are carried over a new gossipsub topic with payload size bounded by `MAX_PROOF_SIZE`. Validation, anti-DoS rate limits, and peer scoring for invalid proofs follow the same patterns as other CL gossip topics; a misbehaving peer is bounded and can be downscored.

**Soundness and consensus implications.** This EIP does not wire proof verification into fork choice or any other consensus rule: `process_execution_proof` runs outside the beacon-block state-transition function. A forged proof therefore cannot fork the chain, slash a validator, or otherwise affect consensus state — its effect is bounded to a verifying node's local view of payload validity.

> **Liveness and critical-path latency.** Proof generation and verification are off the attestation hot path. Validators MUST NOT delay block validation or attestation production while waiting for a proof; if a proof is missing or late, the node attests using the fork choice determined by the execution engine's re-execution of payloads.

Reviewer: Do not use RFC 2119 keywords outside the Specification section. The reason for this is that this might declare new specifications in non-specification sections, which is confusing. To fix, lowercase it.

```suggestion
**Liveness and critical-path latency.** Proof generation and verification are off the attestation hot path. Validators must not delay block validation or attestation production while waiting for a proof; if a proof is missing or late, the node attests using the fork choice determined by the execution engine's re-execution of payloads.
```

eip: 8025
title: Optional Execution Proofs
description: Introducing optional execution proofs on the consensus layer
author: Kevaundray Wedderburn (@kevaundray), Justin Drake (@JustinDrake) <justin@ethereum.org>, Ignacio Hagopian (@jsign), Han (@han0110), Francesco Risitano (@frisitano)
discussions-to: https://ethereum-magicians.org/t/eip-optional-execution-proofs/25500
status: Draft
type: Standards Track
category: Core
created: 2025-09-17
---

## Abstract

> Execution proofs enable consensus layer (CL) nodes to perform stateless validation of execution payloads. Public-good *proof-generating nodes* produce execution proofs and broadcast them over the CL p2p network. *Proof-verifying nodes* consume gossiped execution proofs to assert the validity of execution payloads. Verification is *stateless* and sublinear in the gas limit, so a validator's resource requirements scale sublinearly, an improvement on the linear cost of re-execution. The mechanism is fully opt-in and changes no consensus rules, allowing the proving stack to mature in production before future promotion in mandatory proofs.

Reviewer 2: Nit: maybe `Altruistic` instead of `Public-good`?

## Motivation

> Today, verifying a beacon block requires re-execution of its execution payload over the EL state. The cost of this re-execution scales linearly with the gas limit, and it requires the node to maintain the entire EL state. This couples a node's resource requirements to the state size and the chain's throughput.

Reviewer: Nit: Although clear for EL developers what is meant by state here I still think it is better to be more explicit that this is the entire account and storage trie data. EL state could also be interpeted as ("the EL is ready" (whatever that means)).

Author: Good catch. I updated the first usage of EL state to be "EL state — the full set of account and storage trie data." on line 19. Subsequent usage of EL state does not reference account and storage trie data but happy to change that if you think it's preferable?

This EIP introduces an opt-in path with the following properties:

> - **Sublinear verification cost.** Verifying an execution proof takes approximately constant work, regardless of how much gas the underlying execution payload consumes.

Reviewer: What is approximately constant work? The claim is that it is sublinear, but here the claim is approx constant (so in big O terms O(1), not O(n^c) with c < 1). Abstract claim against gas limit, this is a detail which if correct should be added here. Without context it reads as approx constant time independent of payload parameters e.g. (block) gas limit.

If the proof size scales with the gas limit, but the proof size itself or the verification of big-sized proofs is sublinear (or constant) this should be mentioned here (i.e. introduce more parameters here, for instance block gas limit and/or proof size)

Author: I've gone ahead and added the more accurate big O terms (representing STARKs) as suggested:

> Both verifier cost and proof size scale as `O(log² n)` in the size `n` of the witnessed computation. Since `n` is bounded by the payload's gas limit, verifier cost and proof size are correspondingly sublinear in the gas limit.

It's a little challenging to include the gas limit directly in big O notation, since the gas used is not linear with respect to the computation cycles in the zkEVM. A single gas unit results in a different number of zkEVM cycles depending on which opcode/precompile is invoked. Future gas-repricing based on proving cycles is intended to fix this relationship.

Let me know what you think.

Author: Updated to be:

> Sublinear verification cost. Both verifier runtime and proof size scale as O(log² n) in n, the number of execution steps of the guest program executing the payload. n is bounded by the payload's gas limit.

- **Stateless verification.** Verifying an execution proof requires only the proof itself and a public commitment to the payload — not the EL state.

> - **Enhanced decentralization.** Sublinear, stateless verification keeps node operation tractable for participants with modest hardware and expertise, even as the gas limit and state grow.

Reviewer: Unrelated comment to style, the verification part might become more decentralized (need less resources) then the availability of state (or nodes able to generate these proofs) could then become more centralized if proportionally less nodes have this state. This could be addressed/noted in the Security Considerations or also motivated as part of Rationale.

Author: Good catch. I've updated the "Security Considerations" section to capture this asymmetry:

> **Verifier–prover decentralization asymmetry.** Stateless, sublinear verification lowers the resource floor for validators, but proof generation itself remains demanding: a prover must hold the full EL state used during execution and run a non-trivial proving stack whose cost scales as `O(n log² n)` in the size `n` of the witnessed computation. If the proportion of verifying-only nodes grows faster than the proportion of nodes able to generate proofs, the set of actors who maintain the full EL state and produce proofs may become more concentrated even as overall validator participation broadens.

Let me know what you think. Happy to amend this section or add it as a rationale as you advise.

Reviewer: I think what would make this a bit more specific here is that this is (?) "enhanced verifier decentralization", as now the verification needs less resources. Not sure if enhanced is the "right" word for this. As you pointed out in the security section the verification now needs less resources, so a good word to indicate that, and then specific for validators might help.
I think adding the verifier word is mandatory here, because it does not improve prover decentralization (?)

Reviewer: I got rather convinced this should be made more clear, this should not be addressed in the later sections, but here, as this is  an important point

Author: @jsign added some improvements to the wording here:
> Broader verifier participation. Sublinear, stateless verification lowers the resource floor for proof-verifying validators, keeping payload-validity checks tractable for participants with modest hardware and expertise even as the gas limit and state grow.

> - **Optional by design.** An optional first step lets clients build operational experience, collect performance metrics, and fine-tune the execution proof stack based on real-world findings — without network-wide consensus implications.

Reviewer: The EIP should be written in the sense that once it is Final then once it is read again (say, in 10 years) the interpretation is clear. As of now it is not decided that mandatory proofs happen. I think a more correct wording here is that optional proofs here allow for testing alongside production (thus in the live network) while not harming the network (enshrining untested things in the protocol or slowing the network down significantly) such that in the future proof we could possibly make these proofs mandatory (new EIP, if this EIP already exists then can link to that)

Author: I've tried to reflect your proposal in the updated version. Happy to amend as required.

    - [SSZ serialization schema](https://github.com/ethereum/execution-specs/blob/ec23140720d6a9257a907c470ba1874623bd7b50/src/ethereum/forks/amsterdam/stateless_ssz.py)
    - [Conformance tests](https://github.com/ethereum/execution-specs/tree/ec23140720d6a9257a907c470ba1874623bd7b50/tests/amsterdam/eip8025_optional_proofs)

### Consensus Layer

A node MAY enable either or both of the following modes:

- **Proof-generating mode.** On receiving a valid block, the node requests an execution proof from the proof engine and broadcasts it over a dedicated gossipsub topic.
- **Proof-verifying mode.** Upon receiving an execution proof, the node provides it to the proof engine for verification. If the proof engine returns a `VALID` result, the fork choice is updated. Proof verification does not require EL state.

At initial deployment, opted-in nodes are strongly recommended to continue running an execution node alongside; local re-execution remains the primary source of fork-choice validity and proofs serve as a supplementary check.

### Execution Layer

This EIP introduces new EL types:

> - **`StatelessInput`.** A type that contains the cryptographically verifiable state required to execute a payload. This type is used as private input to the execution proof guest program.

Reviewer: It is unclear to me where this type is used in context of the EL. Is this part of engine API? The reader should not be forced to go to the specs, it should either be illustrated here or pointed to the spec, so the reader can study the code/spec to understand what it does (a high level description to help the reader is always hand here, the text here is fine, but it is not clear in what context this is supposed to be used)

Author: Understood. Myself and @jsign have added substance to both the CL and EL sections to lift the core components into this EIP. Hopefully, this is now more intuitive, but please let me know if some sections are lacking detail and need bolstering to improve comprehension.

## Backwards Compatibility

This EIP is fully opt-in and does not require a hardfork. No consensus rule changes: validators that do not enable either mode see no change to their behaviour, their bandwidth, or their attestation duties. Nodes that opt in additionally subscribe to the proof gossip topic and may run a prover; this affects only their local resource profile.

## Test Cases

- [Execution Layer](https://github.com/ethereum/execution-specs/tree/ec23140720d6a9257a907c470ba1874623bd7b50/tests/amsterdam/eip8025_optional_proofs)
- [Consensus Layer](https://github.com/ethereum/consensus-specs/tree/5aa6eec83af11bb67e7ff69aa647b5736f530342/specs/_features/eip8025)

## Reference Implementation

> - Lighthouse: `https://github.com/eth-act/lighthouse/tree/1480f94c094382ab092362b0b160dc8206acf18f`

Reviewer: Using tricks to pass the EIP bot is not allowed, see EIP-1 for the allowed URLs.

Author: Got it, I've removed the link to the lighthouse repo.

> - [Execution Layer](https://github.com/ethereum/execution-specs/tree/projects/zkevm)

Reviewer 2: While it is conceptually better to have direct links to always reference the latest stuff, I think all external links to GH are checked to be permalinks to avoid abitrary change if the link gets compromised. So maybe we should stick to the same commit used in other permalinks before.

Author: Ahh yes, good catch! Will update.
